test/e2e/util_gatewayapi_test.go

```diff
@@ -590,6 +623,103 @@ func assertGatewaySuccessful(t *testing.T, namespace, name string) (*gatewayapiv
 	return gw, nil
 }
 
+func waitForGatewayListenerCondition(t *testing.T, gatewayName types.NamespacedName, listenerName string, conditions ...metav1.Condition) error {
+	t.Helper()
+
+	gateway := &gatewayapiv1.Gateway{}
+	expected := gatewayListenerConditionMap(conditions...)
+	current := map[string]string{}
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, gatewayName, gateway); err != nil {
+			t.Logf("failed to get gateway %s: %v", gatewayName.Name, err)
+			return false, nil
+		}
+
+		var listenerStatus *gatewayapiv1.ListenerStatus
+		for _, ls := range gateway.Status.Listeners {
+			if string(ls.Name) == listenerName {
+				listenerStatus = &ls
+				current = gatewayListenerConditionMap(listenerStatus.Conditions...)
+				return conditionsMatchExpected(expected, current), nil
+			}
+		}
+
+		if listenerStatus == nil {
+			t.Logf("listener %s not found in gateway %s", listenerName, gatewayName.Name)
+			return false, nil
+		}
+		return false, nil
```

Comment:
If you want to stop here do this:
```suggestion
		if listenerStatus == nil {
			t.Logf("listener %s not found in gateway %s", listenerName, gatewayName.Name)
			return false, fmt.Errorf("listener %s not found in gateway %s", listenerName, gatewayName.Name)
		}
```
If you want to retry, do this:
```suggestion
		if listenerStatus == nil {
			t.Logf("listener %s not found in gateway %s, retrying...", listenerName, gatewayName.Name)
			return false, nil
		}
```
Looks like we don't need the last `return false, nil`.

Reply:
Thank you. The intent was to retry there. Modified the log to reflect that.
Also about the last return, seeing: `Missing the 'return' statement at the end of the function`